File: ruby/lib/codebreaker/marker.rb
module Codebreaker
  class Marker
    def initialize(secret, guess)
      @secret = secret
      @guess = guess
    end

    def number_match_count
      total_match_count - exact_match_count
    end

    def total_match_count
      secret = @secret.chars.to_a
      @guess.chars.to_a.inject(0) do |count, g|
        count + (delete_first(g, secret) ? 1 : 0)
      end
    end

    def delete_first(c, code)
      code.delete_at(code.index(c)) if code.index(c)
    end

    def exact_match_count
      (0..3).inject(0) do |count, index|
        count + (@secret[index] == @guess[index] ? 1 : 0)
      end
    end
  end
end

File: python/lib/marker.py
class Marker(object):
    def __init__(self, secret):
        self.secret = secret

    def exact_match_count(self, guess):
        return reduce (lambda count, i: count + (1 if guess[i] == self.secret[i] else 0), range(0,4), 0)
    
    def number_match_count(self, guess):
        return self.total_match_count(guess) - self.exact_match_count(guess)
    
    def total_match_count(self, guess):
        secret = list(self.secret)
        return reduce(lambda count, g: count + (1 if self.delete_first(g, secret) else 0), guess, 0)

    def delete_first(self, g, secret):
        try:
            secret.remove(g)
            return True
        except Exception:
            return False
